### ISSUE TITLE
YSQL Indexes on Enum based columns ( user_defined_type )

### ISSUE THREAD

**Issue by h-20-s** [yugabyte/yugabyte-db#4014]
Hi, 

I have a relational database schema that I'm looking to test / port over to YB. 

In my schema I define an Enum type and within one of the tables I have a column that uses this Enum type. I then have a composite index across a few columns on that table including the column of my Enum type. I noticed the following error from SQLAlchemy when trying to create the database:

`sqlalchemy.exc.NotSupportedError: (psycopg2.errors.FeatureNotSupported) INDEX on column of type 'user_defined_type' not yet supported`

NB: With SQLAlchemy you can use "non native" enums, (it creates a varchar field with a column constraint) and I can confirm that works. However it would be ideal to use native enums as that's what we're currently running.

**Comment by ndeodhar** [comment]
Thanks for reporting this @h-20-s . We don't yet support user-defined types in key columns (for both primary key and index). Unfortunately, the work to support this is fairly involved and is likely to be done few months down the line.
It'll be great if you can use the non native enums until then.

**Comment by sstubbs** [comment]
I have the same issue trying to install harbor. 
```
{"Event":"pgsql: an error occured while running migrations: migration 5 failed: pq: INDEX on column of type 'user_defined_type' not yet supported","Level":"fatal","Location":"main.go:97","Time":"2020-08-31 12:47:55.566181"}
```

**Comment by rafaelfesi** [comment]
I had the same issue on Elixir with the Phoenix Framework.

**Comment by jaki** [comment]
Looked into this a bit.  The [postgresql docs][1] say

> there is support for adding new values to an existing enum type, and for renaming values

and

> Existing values cannot be removed from an enum type, nor can the sort ordering of such values be changed, short of dropping and re-creating the enum type.

If it weren't for the ability to add new values _at any position_, this would have been straightforward: just store the position as an int64 and don't worry about it ever changing.  So the question is, **do people need this feature to add elements at any position after having created the enum**?  I think that if delete or resort won't be supported, not supporting addition isn't so bad either.  We can still support limited addition, likely adding to the very end, possibly also adding to the very beginning (going to negative numbers).

Adding an element to the enum when the enum is in use for a btree index isn't that hard since the btree structure won't really change even if some elements' sort numbers get bumped up (in the case the addition was to the beginning or middle).  However, since Yugabyte uses DocDB record-based storage, it will be difficult to avoid changing the storage when adding an enum element (to beginning or middle).

We could do some scheme where numbers stored are jumps of, say 2^(64-8) at first so that after 2^8 elements or so are inserted, we do some rebuild operation on the next "add element".  This will be an amortized penalty, but I wonder how we would be able to hunt down all the indexes/tables using the enum as a key (and possibly nonkey ones too) and whether it is even feasible to alter such tables/indexes under the hood.

Otherwise, I don't see another good strategy that avoids having to touch existing storage.  Therefore, I think it is better to not support adding elements to the middle or beginning.

The other concern is backwards compatibility.  I see one problem: current enums use int64 conversion funcs when they should use uint32 (or, more accurately, oid) conversion.  We _could_ leave it at converting to int64: I imagine we can take advantage of this to also support adding elements to the beginning and use negative numbers without worrying of running out of space since it's capped at 2^32.  Then, we avoid backwards compatibility issue of existing enums storage.

We should also try to use the type entity corresponding to `ANYENUMOID` rather than the generic `YBCFixedLenByValTypeEntity`.  When doing that, `ANYENUMOID` needs to change to int64 stuff like `YBCFixedLenByValTypeEntity`.  This should be fine because I believe nothing is using the type entity for `ANYENUMOID` right now.

cc: @kmuthukk, @m-iancu

[1]: https://www.postgresql.org/docs/current/datatype-enum.html

**Comment by skywalkerisnull** [comment]
Is there anywhere that we can follow along with the progress on this particular issue? Like a PR?

We have a few COTS systems that we would like to move across to yugabyte, but on restore we get the following error:

```pg_restore: error: could not execute query: ERROR:  PRIMARY KEY containing column of type 'user_defined_type' not yet supported```

And I am keen to watch for when we can test again with the view to migrating those applications across.

**Comment by jaki** [comment]
@skywalkerisnull, most of our partial work is done on Phabricator, but that is currently internal.  @myang2021, it would be good to provide updates on progress here, like I do in issue #1199, or push incremental code changes to a public fork.

**Comment by myang2021** [comment]
Hi @jaki , I have just started looking into this. I looked at regular postgres behavior and looks like it stores some value that appeared to be a floating point number.

```
create type e_t as enum ('a');
CREATE TYPE
create table foo(id e_t primary key);
CREATE TABLE
insert into foo values ('a');
INSERT 0 1
select * from pg_enum;
 enumtypid | enumsortorder | enumlabel 
-----------+---------------+-----------
     16387 |             1 | a
(1 row)

alter type e_t add value '1' before 'a';
ALTER TYPE
insert into foo values ('1');
INSERT 0 1
alter type e_t add value '0' before 'a';
ALTER TYPE
insert into foo values ('0');
INSERT 0 1
select cast(enumsortorder as float), enumlabel from pg_enum;
 enumsortorder | enumlabel 
---------------+-----------
             1 | a
             0 | 1
           0.5 | 0
(3 rows)

```
So my initial question is why can't we do postgres is doing. I understand that a btree insertion can be more efficient but isn't that argument apply equally to other indexable types such as float, int? In other words, with doc db lsm + YB optimization should be as efficient as a btree right?

**Comment by jaki** [comment]
ohhh, `enumsortorder` is a float.  https://www.postgresql.org/docs/13/catalog-pg-enum.html.  Insertions to the middle and beginning are fine _as long as_ the `enumsortorder` value never changes for an element.  @myang2021, can you check on that?

For example, I wonder what happens if I keep inserting before 'a' until it's no longer possible.

**Comment by myang2021** [comment]
@jaki , thanks for the link, good to know. I did some more testing where I keep "cut in half" until this is no longer possible, then it appears that some kind of remapping/rehashing happens in postgres so internal enumsortorder changes:
```
select cast(enumsortorder as float), enumlabel from pg_enum;
   enumsortorder   | enumlabel 
-------------------+-----------
                 1 | a
                 0 | 24
               0.5 | 23
              0.75 | 22
             0.875 | 21
            0.9375 | 20
           0.96875 | 19
          0.984375 | 18
         0.9921875 | 17
        0.99609375 | 16
       0.998046875 | 15
      0.9990234375 | 14
     0.99951171875 | 13
    0.999755859375 | 12
   0.9998779296875 | 11
  0.99993896484375 | 10
 0.999969482421875 | 9
 0.999984741210938 | 8
 0.999992370605469 | 7
 0.999996185302734 | 6
 0.999998092651367 | 5
 0.999999046325684 | 4
 0.999999523162842 | 3
 0.999999761581421 | 2
  0.99999988079071 | 1
 0.999999940395355 | 0
(26 rows)

yugabyte=# alter type e_t add value '25' before 'a';
alter type e_t add value '25' before 'a';
ALTER TYPE
yugabyte=# select cast(enumsortorder as float), enumlabel from pg_enum;
select cast(enumsortorder as float), enumlabel from pg_enum;
 enumsortorder | enumlabel 
---------------+-----------
            26 | a
            25 | 0
            24 | 1
            23 | 2
            22 | 3
            21 | 4
            20 | 5
            19 | 6
            18 | 7
            17 | 8
            16 | 9
            15 | 10
            14 | 11
            13 | 12
            12 | 13
            11 | 14
            10 | 15
             9 | 16
             8 | 17
             7 | 18
             6 | 19
             5 | 20
             4 | 21
             3 | 22
             2 | 23
             1 | 24
          25.5 | 25
(27 rows)


```
As shown above, after I added one more value, the new internal sort values are updated and there is only one float (25.5) left there. I have tested by keeping "cut-in-half" and this remapping happens several times when needed.

**Comment by m-iancu** [comment]
Fixed by 131ea34.
